File: internal/model/advance_receipt.go
package model

import (
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

type AdvanceReceiptStatus string

const (
	AdvanceReceiptStatusDraft             AdvanceReceiptStatus = "draft"
	AdvanceReceiptStatusConfirmed         AdvanceReceiptStatus = "confirmed"
	AdvanceReceiptStatusPartiallyAllocated AdvanceReceiptStatus = "partially_allocated"
	AdvanceReceiptStatusFullyAllocated    AdvanceReceiptStatus = "fully_allocated"
	AdvanceReceiptStatusReversed          AdvanceReceiptStatus = "reversed"
)

type AdvanceReceipt struct {
	ID               uuid.UUID       `db:"id"`
	TenantID         uuid.UUID       `db:"tenant_id"`
	CompanyID        uuid.UUID       `db:"company_id"`
	CustomerID       uuid.UUID       `db:"customer_id"`
	AdvanceNo        string          `db:"advance_no"`
	Amount           decimal.Decimal `db:"amount"`
	AllocatedAmount  decimal.Decimal `db:"allocated_amount"`
	OutstandingAmount decimal.Decimal `db:"outstanding_amount"`
	ReceivedDate     time.Time       `db:"received_date"`
	DueDate          *time.Time      `db:"due_date"`
	Status           string          `db:"status"`
	SourceType       string          `db:"source_type"`
	BankAccountID    *uuid.UUID      `db:"bank_account_id"`
	ReferenceNo      *string         `db:"reference_no"`
	Remark           *string         `db:"remark"`
	VoucherID        *uuid.UUID      `db:"voucher_id"`
	VoucherNo        *string         `db:"voucher_no"`
	CreatedBy        *uuid.UUID      `db:"created_by"`
	CreatedAt        time.Time       `db:"created_at"`
	ConfirmedBy      *uuid.UUID      `db:"confirmed_by"`
	ConfirmedAt      *time.Time      `db:"confirmed_at"`
	ReversedBy       *uuid.UUID      `db:"reversed_by"`
	ReversedAt       *time.Time      `db:"reversed_at"`
}

File: frontend/src/api/modules/advance_receipt.ts
import request from '@/api/request'
import type { ApiResponse } from '@/types/api'

export interface AdvanceReceipt {
  id: string
  advance_no: string
  customer_id: string
  amount: string
  allocated_amount: string
  outstanding_amount: string
  received_date: string
  due_date?: string
  status: string
  source_type: string
  bank_account_id?: string
  reference_no?: string
  remark?: string
  voucher_no?: string
  created_at: string
  confirmed_at?: string
}

export interface CreateAdvanceReceiptPayload {
  company_id: string
  customer_id: string
  amount: string
  received_date: string
  due_date?: string
  bank_account_id?: string
  reference_no?: string
  remark?: string
}

export function fetchAdvanceReceipts(params: { status?: string } = {}): Promise<ApiResponse<{ list: AdvanceReceipt[]; total: number }>> {
  return request.get('/advance-receipts', { params })
}

export function fetchAdvanceReceiptById(id: string): Promise<ApiResponse<AdvanceReceipt>> {
  return request.get(`/advance-receipts/${id}`)
}

export function fetchOutstandingAdvanceReceipts(customerId: string): Promise<ApiResponse<{ list: AdvanceReceipt[]; total: number }>> {
  return request.get('/advance-receipts/outstanding', { params: { customer_id: customerId } })
}

export function createAdvanceReceipt(payload: CreateAdvanceReceiptPayload): Promise<ApiResponse<AdvanceReceipt>> {
  return request.post('/advance-receipts', payload)
}

export function confirmAdvanceReceipt(id: string): Promise<ApiResponse<AdvanceReceipt>> {
  return request.post(`/advance-receipts/${id}/confirm`)
}
